Expose "Coordinates" as part of Xarray's public API

benbovy opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #7214
- [x] Closes #6392
- [x] xref #6633
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

This is a rework of #7214. It follows the suggestions made in https://github.com/pydata/xarray/pull/7214#issuecomment-1295283938, https://github.com/pydata/xarray/pull/7214#issuecomment-1297046405 and https://github.com/pydata/xarray/pull/7214#issuecomment-1293774799:

- No `indexes` argument is added to `Dataset.__init__`, and the `indexes` argument of `DataArray.__init__` is kept private (i.e., valid only if fastpath=True)
- When a `Coordinates` object is passed to a new Dataset or DataArray via the `coords` argument, both coordinate variables and indexes are copied/extracted and added to the new object
- This PR also adds ~~an `IndexedCoordinates` subclass~~ `Coordinates` public constructors used to create Xarray coordinates and indexes from non-Xarray objects. For example, the `Coordinates.from_pandas_multiindex()` class method creates a new set of index and coordinates from an existing `pd.MultiIndex`.

EDIT: `IndexCoordinates` has been merged with `Coordinates`

EDIT2: it ended up as a pretty big refactor with the promotion of `Coordinates` has a 2nd-class Xarray container that supports alignment like Dataset and DataArray. It is still quite advanced API, useful for passing coordinate variables and indexes around. Internally, `Coordinates` objects are still "virtual" containers (i.e., proxies for coordinate variables and indexes stored in their corresponding DataArray or Dataset objects). For now, a "stand-alone" `Coordinates` object created from scratch wraps a Dataset with no data variables.

Some examples of usage:

```python
import pandas as pd
import xarray as xr

midx = pd.MultiIndex.from_product([["a", "b"], [1, 2]], names=("one", "two"))

coords = xr.Coordinates.from_pandas_multiindex(midx, "x")
# Coordinates:
#   * x        (x) object MultiIndex
#   * one      (x) object 'a' 'a' 'b' 'b'
#   * two      (x) int64 1 2 1 2

ds = xr.Dataset(coords=coords)
# <xarray.Dataset>
# Dimensions:  (x: 4)
# Coordinates:
#   * x        (x) object MultiIndex
#   * one      (x) object 'a' 'a' 'b' 'b'
#   * two      (x) int64 1 2 1 2
# Data variables:
#     *empty*

ds_to_be_deprecated = xr.Dataset(coords={"x": midx})
ds_to_be_deprecated.identical(ds)
# True

da = xr.DataArray([1, 2, 3, 4], dims="x", coords=ds.coords)
# <xarray.DataArray (x: 4)>
# array([1, 2, 3, 4])
# Coordinates:
#   * x        (x) object MultiIndex
#   * one      (x) object 'a' 'a' 'b' 'b'
#   * two      (x) int64 1 2 1 2
```

TODO:

- [x] update `assign_coords` too so it has the same behavior if a `Coordinates` object is passed?
- [x] How to avoid building any default index? It seems silly to add or use the `indexes` argument just for that purpose? ~~We could address that later.~~ Solution: wrap the coordinates dict in a Coordinates objects, e.g., `ds = xr.Dataset(coords=xr.Coordinates(coords_dict))`.

@shoyer, @dcherian, anyone -- what do you think about the approach proposed here? I'd like to check that with you before going further with tests, docs, etc.

benbovy commented:
`IndexedCoordinates` and `Indexes` have a lot of overlap. At some point we might consider merging the two classes, like @shoyer suggests in https://github.com/pydata/xarray/pull/7214#issuecomment-1295283938. The main difference is that one is a mapping of coordinates and the other is a mapping of indexes. `IndexedCoordinates` is mostly reusing `Indexes` and `Dataset` under the hood, it is only a facade.

Alternatively to an `IndexedCoordinates` subclasses I was wondering if we could reuse the `Coordinates` base class? There's some benefit of providing a subclass:

- besides specific constructors like `.from_pandas_multiindex()` it has a generic `__init__` for advanced use cases. Not sure it is a good idea to add this constructor to the base class?
- unlike Coordinates, IndexedCoordinates is immutable.

What if the `Indexes` class was a facade based on `IndexedCoordinates` instead of the other way around? It would probably make more sense but it would also be a bigger refactor. I've chosen the easy way :).

benbovy commented:
I added `IndexedCoordinates.merge_coords` so that it is easier to combine different coordinates to pass to a new Dataset / DataArray, e.g.,

```python
midx = pd.MultiIndex.from_product([["a", "b"], [1, 2]], names=("one", "two"))

coords = xr.IndexedCoordinates.from_pandas_multiindex(midx, "x")

coords = coords.merge_coords({"y": [0, 1, 2]})
# Coordinates:
#   * x        (x) object MultiIndex
#   * one      (x) object 'a' 'a' 'b' 'b'
#   * two      (x) int64 1 2 1 2
#   * y        (y) int64 0 1 2

ds = xr.Dataset(coords=coords)
# <xarray.Dataset>
# Dimensions:  (x: 4)
# Coordinates:
#   * x        (x) object MultiIndex
#   * one      (x) object 'a' 'a' 'b' 'b'
#   * two      (x) int64 1 2 1 2
#   * y        (y) int64 0 1 2
# Data variables:
#     *empty*
```

`IndexedCoordinates.merge_coords` is very much like `Coordinates.merge` except that it returns a new Coordinates object instead of a Dataset.

Or should we just use `merge`? It would require that:

- `Coordinates.merge` accepts `Mapping[Any, Any]` for its `other` argument. Only changing the type hint is enough here since the implementation already accepts any input passed to Dataset.
-  When a Dataset is passed as `coords` argument to a new Dataset and DataArray, both variables and indexes should be extracted. It is already the case for Dataset but only for pandas indexes (default indexes & backwards compatibility).

shoyer commented:
> what do you think about the approach proposed here? I'd like to check that with you before going further with tests, docs, etc.

Generally this looks great to me!

> * How to avoid building any default index? It seems silly to add or use the `indexes` argument just for that purpose? We could address that later.

My suggestion would be:

- `coords` passed as a `dict`: create default indexes
- `coords` passed as `IndexedCoordinates`: do not create defaults

> Alternatively to an IndexedCoordinates subclass I was wondering if we could reuse the Coordinates base class?

Yes, this makes more sense to me!

> What if the Indexes class was a facade based on IndexedCoordinates instead of the other way around? 

Yes, I also agree! This makes more sense.

shoyer commented:
Long term, do you think it would make sense to merge together Indexes, Coordinates and IndexedCoordinates? They are sort of all containers for the same thing.

benbovy commented:
> Long term, do you think it would make sense to merge together Indexes, Coordinates and IndexedCoordinates? They are sort of all containers for the same thing.

Yes I think so.

I'm actually trying to merge `IndexedCoordinates` with `Coordinates` but I'm stuck: the latter is abstract and I don't really see how I could refactor it together with `DatasetCoordinates` and `DataArrayCoordinates`. Do you have any idea on how best to proceed?

Ideally, I'd see `Coordinates` be exposed in Xarray's main namespace with at least the two following constructors:

```python
class Coordinates:

    def __init__(
        self,
        coords: Mapping[Any, Any] | None = None,
        indexes: Mapping[Any, Index] | None = None,
    ):
        # Similar to Dataset.__init__ but without the need
        # to merge coords and data vars...
        # Probably ok to allow more flexibility / less safety here?
        ...

    @classmethod
    from_pandas_multiindex(cls, index: pd.MultiIndex, dim: str):
        ...
```

shoyer commented:
> I'm actually trying to merge `IndexedCoordinates` with `Coordinates` but I'm stuck: the latter is abstract and I don't really see how I could refactor it together with `DatasetCoordinates` and `DataArrayCoordinates`

`Coordinates` is abstract because in the (current) Xarray data model, it doesn't actually store any data -- coordinates are stored in private attributes of the original `Dataset` (`._variables` and `._coord_names`) or `DataArray` (`._coords`). So `Coordinates` needs to serve as a proxy for the data.

In the long term, I think we should refactor `Dataset`/`DataArray` to actually store data (coordinate variables, indexes and dimension sizes) on `Coordinates`, but that's a bigger refactor.

For now, it's worth noting that the current `Coordinates` class isn't actually exposed in Xarray's public API, just the `DatasetCoordinates` and `DataArrayCoordinates` classes (and not even their constructors). So the intermdiate step I would try is:
1. Rename the current `Coordinates` baseclass to `AbstractCoordinates`.
2. Rename your `IndexedCoordinates` class to `Coordinates`. Expose it in the public API. Make sure it can handle `DatasetCoordinates` and `DataArrayCoordinates` in the constructor.
3. Maybe: use some Python magic to make DatasetCorodinates/DataArrayCoordinates subclasses of the new Coordinates. Or maybe make them actual subclassses, overriding many of the methods (including the constructor).

benbovy commented:
Thanks @shoyer, I've been thinking about similar short/long term plans although so far I haven't figured out how to implement your point 3. I'll give it another try.

benbovy commented:
> My suggestion would be:
    coords passed as a dict: create default indexes
    coords passed as IndexedCoordinates: do not create defaults

So if we already have some coordinate data as a dict but don't want any default index, we would need to do this:

```python
ds = xr.Dataset(coords=xr.Coordinates(my_coord_dict))
```

instead of this:

```python
ds = xr.Dataset(coords=my_coord_dict)
```

benbovy commented:
I think this is ready for review!

benbovy commented:
I did some profiling to find the cause of the decrease in performance reported in the benchmarks (dataset creation). In summary, this is explained by a `Coordinates` object (built from the `coords` mapping) that is now included in objects to align when merging data vars and coordinates. Previously all non DataArray objects in the `coords` mapping were excluded from alignment (in `deep_align`). The introduced overhead comes from a call to `Coordinates._reindex_callback()`, which (I think?) should do no more than shallow copies and/or xarray wrapping stuff. In the benchmark report this is only marked as significant when creating small datasets (1.5-2x slower), and it becomes insignificant for datasets with more data variables.

Maybe there's some way to optimize that? I don't know if we can completely avoid it with the solution implemented in this PR, though. Promoting `Coordinates` is pretty clean and future proof IMO (assuming that we'll further refactor `Coordinates` to actually store variables and indexes, i.e., not as a proxy anymore). Is the (minor? temporary?) regression in performance acceptable and can we just leave like that?

More details about the new workflow implemented in this PR when creating a new Dataset:

- if Dataset's `coords` argument is a "simple" mapping, it is first internally converted into a `Coordinates` object, with the creation of default indexes for dimension coordinates
  - if one or more DataArray objects are given in `coords`, their coordinates (variables + indexes) are extracted and merged with the other input coordinates
  - see the implementation in `xarray.core.coordinates.create_coords_with_default_indexes`
- otherwise, just reuse the `Coordinates` object passed as `coords`
- coordinates are then merged with data variables
  - the `Coordinates` object is aligned with every other "alignable" object found in `data_vars`
  - coordinate indexes (if any) are passed explicitly to `align` so they are used in priority
  - explicitly using a `Coordinates` object skips the creation of default indexes during merging (in `collect_variables_and_indexes()`)

benbovy commented:
> Maybe there's some way to optimize that? I don't know if we can completely avoid it with the solution implemented in this PR, though. Promoting Coordinates is pretty clean and future proof IMO (assuming that we'll further refactor Coordinates to actually store variables and indexes, i.e., not as a proxy anymore). Is the (minor? temporary?) regression in performance acceptable and can we just leave it like that for now?

Fixed in [193dad3](https://github.com/pydata/xarray/pull/7368/commits/193dad3393565b6c007c0eb0a2d47b5ade874571) (with some reasonable special case added in `merge_core`).

shoyer commented on xarray/core/coordinates.py:
```diff
@@ -241,13 +456,115 @@ def merge(self, other: Coordinates | None) -> Dataset:
             variables=coords, coord_names=coord_names, indexes=indexes
         )
 
+    def merge_coords(self, other: Mapping[Any, Any] | None = None) -> Coordinates:
```
I'm not sure this name is worth the trouble, despite the better API. `coords.merge(other).coords` is at least as succinct as `coords.merge_coords(other)`.

Instead, we could consider a deprecation cycle to remove the existing `.merge()`, or perhaps even change the return type from `Dataset` to `Coordinates`? That would be a breaking change, but also one with very low impact. There are very few uses of `Coordinates.merge` currently and all of those could be replaced by calling `xarray.merge()`.

benbovy replied:
Agreed.

I added `merge_coords` before adding support for `Coordinates` objects in `xarray.merge()` and `xarray.align()`. It's better to reuse those instead of a new method.

> or perhaps even change the return type from Dataset to Coordinates?

Yes that would make sense.

This makes me also think about what to do when a Dataset object is passed as `coords` argument? I think that currently it is handled like a regular mapping (indexes are ignored). Is it what we want or should we treat it like `Coordinates` and extract indexes too?

shoyer commented on xarray/core/coordinates.py:
```diff
@@ -482,3 +800,60 @@ def assert_coordinate_consistent(
+    indexes: dict[Hashable, Index] = {}
+    variables: dict[Hashable, Variable] = {}
+
+    # this is needed for backward compatibility: when a pandas multi-index
+    # is given as data variable, it is promoted as index / level coordinates
```
Can we issue a FutureWarning in this case, to note that MultiIndex should be supplied in coordinates?

benbovy replied:
I was thinking adding it in a follow-up PR (as well as for any MultiIndex given as coordinate data), but yes we can do it here.

shoyer commented on xarray/core/dataarray.py:
```diff
@@ -1789,7 +1817,11 @@ def _reindex_callback(
-        return self._from_temp_dataset(reindexed)
+
+        da = self._from_temp_dataset(reindexed)
+        da.encoding = self.encoding
```
is this fixing something else? a test would not be a bad idea :)

benbovy replied:
It is not fixing anything specific here, it has been moved from `Aligner` into the Dataset / DataArray callbacks when adding alignment support for Coordinates. I guess it is already tested somewhere?

shoyer commented on xarray/core/merge.py:
```diff
@@ -748,10 +707,20 @@ def merge_core(
 
     _assert_compat_valid(compat)
 
+    objects = list(objects)
+    if skip_align_args is None:
+        skip_align_args = []
+
+    skip_align_objs = [(pos, objects.pop(pos)) for pos in skip_align_args]
+
     coerced = coerce_pandas_values(objects)
     aligned = deep_align(
         coerced, join=join, copy=False, indexes=indexes, fill_value=fill_value
     )
+
+    for pos, obj in skip_align_objs:
+        aligned.insert(pos, obj)
```
Could we equivalently just not call `merge_core` with these skipped arguments? What does it mean to include a skipped argument here?

benbovy replied:
It is used when merging `coords` and `data_vars` during the creation of a new Dataset:

`coords` has already been coerced to a `Coordinates` object and we want to exclude it from alignment (coordinate variables should already be aligned together and it shouldn't be re-indexed). But we still need to merge it with the `data_vars` mapping and align data variables with coordinate indexes.

shoyer commented on xarray/testing.py:
```diff
@@ -78,11 +79,16 @@ def assert_equal(a, b):
     """
     __tracebackhide__ = True
-    assert type(a) == type(b)
+    try:
+        assert type(a) == type(b)
+    except AssertionError:
+        assert isinstance(a, Coordinates) and isinstance(b, Coordinates)
```
```suggestion
    assert (
        type(a) == type(b) or
        isinstance(a, Coordinates) and isinstance(b, Coordinates)
    )
```

shoyer commented on xarray/testing.py:
```diff
@@ -96,24 +102,29 @@ def assert_identical(a, b):
     """
     __tracebackhide__ = True
-    assert type(a) == type(b)
+    try:
+        assert type(a) == type(b)
+    except AssertionError:
+        assert isinstance(a, Coordinates) and isinstance(b, Coordinates)
```
Same as above

shoyer commented on xarray/tests/test_coordinates.py:
```diff
@@ -0,0 +1,141 @@
+        for name in ("x", "one", "two"):
+            assert_identical(expected[name], coords.variables[name])
+
+    def test_dims(self, coords) -> None:
```
I would not use `coords` as a fixture on all these methods. Instead, I would re-create `coords` explicitly in each test.

The general guidance is only use generic fixtures/setup for creating objects if all somebody reading the code needs to know is that it's _some_ generic coordinates object, e.g., this would be OK:
```
def test_dataset_coords(self, coords):
  assert Dataset(coord).coords == coords
```

But here you're testing for particular values. For cases like this, it's better to make tests self-contained, even if they are a little redundant.

shoyer reviewed (commented):
Generally looks great!

benbovy commented:
Thanks for the review @shoyer. I addressed your comments.

Everything seems OK except a rather annoying mypy error that I'm struggling with:

The `DataAlignable` type variable should now encompass both `DataWithCoords` and `Coordinates`, since in this PR we add alignment support for the latter. I somewhat naively tried the options below without success:

- `DataAlignable = TypeVar("DataAlignable", bound=DataWithCoords | Coordinates)` -> doesn't work since we cannot mix DataWithCoords and Coordinates when aligning each object (input type = output type)
- `DataAlignable = TypeVar("DataAlignable", bound=DataWithCoords, Coordinates)` -> doesn't work with subclasses
- `DataAlignable = TypeVar("DataAlignable", Dataset, DataArray, Coordinates)` -> doesn't work with generic types `T_Dataset`, etc.?
- I even tried using a Protocol

@headtr1ck @Illviljan any idea?

headtr1ck commented:
> DataAlignable = TypeVar("DataAlignable", bound=DataWithCoords | Coordinates) -> doesn't work since we cannot mix DataWithCoords and Coordinates when aligning each object (input type = output type)

I think thats exactly correct.
The whold idea of using a TypeVar for the inputs means that all inputs must have the same type.

Consider the following example:
```python
class A: ...
class B(A): ...
class C: ...

T = TypeVar("T", bound=Union[A, C])
def f(*x: T) -> T: ...

f(A(), A())  # OK
f(A(), B())  # OK
f(C(), C())  # OK
f(A(), C())  # not ok, because the common type of A and C is object
```

dcherian commented:
Is this just stuck on a mypy error?

dcherian commented:
@headtr1ck / @Illviljan Do you have any ideas for how to type this?

> The DataAlignable type variable should now encompass both DataWithCoords and Coordinates, since in this PR we add alignment support for the latter.

Illviljan commented:
```
xarray/tests/test_coordinates.py:139: error: Value of type variable "DataAlignable" of "align" cannot be "Coordinates"  [type-var]
xarray/tests/test_coordinates.py:144: error: Value of type variable "DataAlignable" of "align" cannot be "Coordinates"  [type-var]

Found 2 errors in 1 file (checked 142 source files)
```

headtr1ck commented:
> @headtr1ck / @Illviljan Do you have any ideas for how to type this?
> 
> > The DataAlignable type variable should now encompass both DataWithCoords and Coordinates, since in this PR we add alignment support for the latter.

I don't think we need to keep the generalization of DataWithCoords. We could just keep it at a typevar that is either DataArray, Dataset or Coords.

The only problem is that when you have functions like `f(*args)` the args type becomes the common ancestor of all Inputs, which for DataArray and Dataset was DataWithCoords. But if you now add a Coordinates object it will default to a python object type (or maybe some common mixin class).

I dont think there is any way around that.
We could try adding a few overloads for 1-3 arguments which is the most common usecase, but then you'll always have this problem with >3 args.

headtr1ck commented:
It seems that it is not that easy sadly.
Alignment uses a lot of methods that are defined only at `DataArray`, `Dataset` and `Coordinate` level.
So `DataWithCoords` is actually not allowed. I have to see if I can resolve these conflicts...

dcherian commented:
@benbovy Seems like we need to add a `coords` argument to `open_dataset` to finally fix #2233. Is that right?

dcherian commented on xarray/__init__.py:
```diff
@@ -26,6 +26,7 @@
     where,
 )
 from xarray.core.concat import concat
+from xarray.core.coordinates import Coordinates
```
Turns out there is a `Coordinate` top-level import from Variable. We should delete that to avoid confusion:

https://github.com/pydata/xarray/blob/a47ff4ed7ff6d6a5a16d0e604bdb120557c42e13/xarray/__init__.py#L38
https://github.com/pydata/xarray/blob/a47ff4ed7ff6d6a5a16d0e604bdb120557c42e13/xarray/core/variable.py#L3142

benbovy replied:
Should we go ahead and delete it without deprecation notice? I'm not sure if anyone is still using `Coordinate`, though, this is quite old.

dcherian replied:
IMO it's ok to delete. It's been so long...

benbovy replied:
Done!

headtr1ck commented on xarray/core/types.py:
```diff
@@ -43,18 +47,15 @@
+    import sys
+
+    if sys.version_info >= (3, 11):
+        from typing import Self
```
Nice! Finally this is merged.

This means we should be able to remove lots of the code that annotated self with a typevar (in another PR ofc).

headtr1ck commented on xarray/core/types.py:
```diff
@@ -93,10 +94,63 @@
     DTypeLikeSave: Any = None
 
 
+class Alignable(Protocol):
```
Nice!

headtr1ck reviewed (approved):
I totally forgot that you can use Protocols to solve this typing issue, nice!

benbovy commented:
Thanks for your feedback @headtr1ck, your previous comments have been helpful. Yeah I gave it another try using `Protocol`, I've been struggling with it until trying with `Self`. This should be good now, hopefully. Another advantage is that the `Alignable` protocol class makes it pretty clear what is the required interface for an Xarray object to support alignment, i.e., it provides some self-documentation of the internals.

headtr1ck commented on xarray/core/coordinates.py:
```diff
@@ -163,13 +179,209 @@ def to_index(self, ordered_dims: Sequence[Hashable] | None = None) -> pd.Index:
+
+    @classmethod
+    def from_pandas_multiindex(
+        cls: type[T_Coordinates], midx: pd.MultiIndex, dim: str
```
And here?

headtr1ck commented on xarray/core/coordinates.py:
```diff
@@ -236,13 +449,86 @@ def merge(self, other: Coordinates | None) -> Dataset:
+        return cast(T_Coordinates, results.coords)
+
+    def _reindex_callback(
+        self: T_Coordinates,
```
And here (not sure if it works with cast though..., Can always use a type: ignore)

headtr1ck commented on xarray/core/coordinates.py:
```diff
@@ -236,13 +449,86 @@ def merge(self, other: Coordinates | None) -> Dataset:
+            exclude_dims,
+            exclude_vars,
+        )
+        return cast(T_Coordinates, aligned.coords)
```
I have not checked this, but maybe mypy is right to complain?

If you start with a DataArrayCoordinates, do you not end up with a DatasetCoordinates?

benbovy replied:
> If you start with a DataArrayCoordinates, do you not end up with a DatasetCoordinates?

This case shouldn't happen because Dataset and DataArray have their own `_reindex_callback()` called internally in `Aligner`. The callback here is called when aligning a `Coordinates` object, so we need to cast `DatasetCoordinates` into a `Coordinates`.

I should probably add a comment here to explain this.

benbovy commented on xarray/core/coordinates.py:
```diff
@@ -163,13 +179,209 @@ def to_index(self, ordered_dims: Sequence[Hashable] | None = None) -> pd.Index:
+
+    @classmethod
+    def _construct_direct(
+        cls: type[T_Coordinates],
```
Yes 👍  I'll go trough the coordinate classes and see if I can replace type variables by Self.

dcherian commented:
RTD failure is real:

```
home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/7368/doc/api-hidden.rst:9: WARNING: autosummary: failed to import Coordinates.merge_coords.
Possible hints:
* ModuleNotFoundError: No module named 'xarray.Coordinates'
* PycodeError: no source found for module 'builtins'
* AttributeError: type object 'Coordinates' has no attribute 'merge_coords'
* ModuleNotFoundError: No module named 'Coordinates'
* KeyError: 'Coordinates'
/home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/7368/doc/api-hidden.rst:9: WARNING: autosummary: failed to import core.coordinates.DatasetCoordinates.merge_coords.
Possible hints:
* ModuleNotFoundError: No module named 'xarray.core.coordinates.DatasetCoordinates'; 'xarray.core.coordinates' is not a package
* PycodeError: no source found for module 'builtins'
* AttributeError: type object 'DatasetCoordinates' has no attribute 'merge_coords'
* KeyError: 'core'
* ModuleNotFoundError: No module named 'core'
/home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/7368/doc/api-hidden.rst:9: WARNING: autosummary: failed to import core.coordinates.DataArrayCoordinates.merge_coords.
Possible hints:
* PycodeError: no source found for module 'builtins'
* KeyError: 'core'
* AttributeError: type object 'DataArrayCoordinates' has no attribute 'merge_coords'
* ModuleNotFoundError: No module named 'xarray.core.coordinates.DataArrayCoordinates'; 'xarray.core.coordinates' is not a package
* ModuleNotFoundError: No module named 'core'
```

benbovy commented:
@dcherian yes this PR would enable #6633 by passing a `Coordinates` object to the Dataset constructor in the backend code. It coud be done in another PR indeed.

- This doesn't create a default index for "x" (since no index is passed explicitly to the `Coordinates` object):

```python
import xarray as xr

ds = xr.Dataset(
    data_vars={"foo": ("x", [0.0, 1.0])},
    coords=xr.Coordinates({"x": ("x", [1, 2])}),
)
```

- Note that if "x" is passed  as a data variable it will still be promoted as a coordinate with a default index, even when a `Coordinates` object is passed explicitly:

```python
xr.Dataset(
    data_vars={"foo": ("x", [1.0, 2.0]), "x": [1, 2]},
    coords=xr.Coordinates(),
)
```

This might be slightly confusing but I don't know if it is really worth handling this special case (i.e., coordinate variables in data_vars + empty Coordinates). I think we should keep promoting data variables as coordinates since `data_vars` is often used as a unique positional argument of `Dataset.__init__()`.

benbovy commented:
I updated the docstrings of `Dataset.__init__` and `DataArray.__init__` to clarify that we can pass a `Coordinates` object to bypass the creation of default indexes for dimension coordinates.

I took this opportunity to improve the glossary in the documentation and make the concepts of (non)-dimension / (non)-indexed coordinate more in line with the current state of Xarray since the index refactor. I'm sure there are still other inconsistent parts in the documentation, though. I'll review and address that in a follow-up PR.

TomNicholas commented:
I just want to say this is another amazing piece of work @benbovy :clap: :clap:

dcherian commented:
@benbovy shall we merge?

benbovy commented:
@dcherian yes!
